### ISSUE TITLE
🔌 Enable asynchronous plugins

### ISSUE THREAD

**Issue by markbrouch** [pugjs/pug#2838]
**Pug Version:** `v2.0.0-rc.2`

**Node Vesrion:** `v8.0.0`

Currently, the Pug plugin system is synchronous, in that each plugin is applied immediately after the previous. I am proposing an asynchronous system for plugins so that each step in `compileBody` is halted until an optional callback function is called by each applied plugin (default callback would be invoked synchronously).

This would allow for plugins to handle async activity, like network or filesystem activity. For my own use, I'd like to be able to implement asynchronous postcss plugins like postcss-modules in a Pug plugin.

**Comment by ForbesLindesay** [comment]
This may be something we do some day, but it won't be soon, and it would probably be limited to the `pre`/`post` plugins, not the lexer/parser/linker/code-gen plugins.  In the mean time, the core pug package is actually pretty simple now, so you can always rebuild the building blocks with whatever async magic you need thrown in.

**Comment by markbrouch** [comment]
Good idea on reimplementing the pug core custom - I might end up going down that route. Unfortunately, that would mean I'd also have to ditch pug-cli, but seems like a good solution overall.

Agreed that it only makes sense for the `pre`/`post` plugins. Would you entertain a PR if I worked up the right appetite, or is this a bit too big of a core change for the 2.0 release?

**Comment by ForbesLindesay** [comment]
I don't want to consider this for the 2.0 release, so it's likely going to be 3.0.0 even if you submit a pull request.

**Comment by markbrouch** [comment]
Fair enough - thanks for the insight.
